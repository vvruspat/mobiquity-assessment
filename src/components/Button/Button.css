.button {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    gap: var(--padding-small);
    height: 40px;
    padding: 0 var(--padding-form-item);
    box-sizing: border-box;
    border-radius: 6px;
    cursor: pointer;
    opacity: 1;
    transition: all 0.1s ease-in;
  
    font-size: 15px;
    line-height: 24px;
    font-weight: bold;
  }
  
  .primary-button {
    background-color: var(--primary-background);
    color: var(--primary-text);
    border: none;
  }
  
  .secondary-button {
    background-color: var(--secondary-background);
    color: var(--secondary-text);
    border: 1px solid var(--border-color);
  }
  
  .outline-button {
    background-color: var(--background);
    color: var(--text);
    border: 1px solid var(--border-color);
    font-weight: 500;
  }
  
  .icon-button {
    background-color: var(--background);
    color: var(--text);
    border: 0 none;
    font-weight: 500;
  }
  
  .stretched-button {
    display: block;
    width: 100%;
  }
  
  .button:active {
    box-shadow: 0 0 1px;
    opacity: 0.8;
    transition: all 0.1s ease-in;
  }
  
  .button:disabled {
    opacity: 0.6;
    cursor: initial;
  }
  
  .button-before,
  .button-after {
    display: flex;
    flex-direction: column;
    justify-content: center;
  }