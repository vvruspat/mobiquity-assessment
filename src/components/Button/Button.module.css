.button {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    height: 36px;
    padding: var(--padding-regular) var(--padding-large);
    box-sizing: border-box;
    border-radius: 30px;
    cursor: pointer;
    transition: all 0.1s ease-in;
  }
  
  .primary-button {
    background-color: var(--regular-color);
    color: var(--regular-inverted-color);
    border: 1px solid var(--regular-color);
  }
  
  .outline-button {
    background-color: transparent;
    color: var(--regular-color);
    border: 1px solid var(--regular-color);
    font-weight: 500;
  }
  
  /* .icon-button {
    background-color: var(--background);
    color: var(--text);
    border: 0 none;
    font-weight: 500;
  } */
  
  .button:hover {
    opacity: 0.9;
    transition: all 0.1s ease-in;
  }

  .button:active {
    opacity: 0.8;
    transition: all 0.1s ease-in;
  }
