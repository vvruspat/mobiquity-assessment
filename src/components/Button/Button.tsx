import cn from "clsx";
import s from "./Button.module.css";

type ButtonProps = React.DetailedHTMLProps<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
> & {
  /** Button style look. */
  mode?: "primary" | "outline" | "icon";
};

/** 
 Styled HTML button with additional functionality
 */
export const Button = ({
  className,
  mode = "primary",
  children,
  ...buttonProps
}: ButtonProps) => {
  return (
    <button
      className={cn(s.button, s[`${mode}-button`], className)}
      {...buttonProps}
    >
      {children}
    </button>
  );
};
