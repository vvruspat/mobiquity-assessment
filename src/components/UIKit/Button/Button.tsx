import cn from "clsx";
import { useCallback, useState } from "react";
import s from "./Button.module.css";

type ButtonProps = React.DetailedHTMLProps<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
> & {
  /** Button style look. */
  mode?: "primary" | "outline" | "icon";
  /**
   * Switch on button rotation on click
   * @description Appliable only for icon button
   * */
  isToggleable?: boolean;
};

/** 
 Styled HTML button with additional functionality
 */
export const Button = ({
  className,
  mode = "primary",
  onClick,
  children,
  isToggleable,
  ...buttonProps
}: ButtonProps) => {
  const [isToggle, setIsToggle] = useState(false);

  const onButtonClick = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      if (isToggleable) {
        setIsToggle((prevToggle) => !prevToggle);
      }

      onClick?.(event);
    },
    [isToggleable, onClick]
  );

  return (
    <button
      className={cn(
        s.button,
        s[`${mode}-button`],
        { [s.toggle]: isToggle },
        className
      )}
      onClick={onButtonClick}
      {...buttonProps}
      data-testid="button"
    >
      {children}
    </button>
  );
};
