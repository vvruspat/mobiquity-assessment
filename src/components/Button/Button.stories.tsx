import { ComponentStory, ComponentMeta } from "@storybook/react";

import { Button } from "./Button";

export default {
  title: "UI Kit/Button",
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => (
  <Button {...args}>Button</Button>
);

export const Primary = Template.bind({});
Primary.args = {};

export const Secondary = Template.bind({});
Secondary.args = {
  mode: "secondary",
};

export const Outline = Template.bind({});
Outline.args = {
  mode: "outline",
};

export const Stretched = Template.bind({});
Stretched.args = {
  stretched: true,
};
