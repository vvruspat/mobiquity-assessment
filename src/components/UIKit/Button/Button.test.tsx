import { render, screen, fireEvent } from "@testing-library/react";
import { Button } from "./Button";

const onClick = jest.fn();

test("renders button", () => {
  render(<Button onClick={onClick}>Test button</Button>);

  const el = screen.getByTestId("button");

  fireEvent.click(el);
  expect(onClick).toHaveBeenCalledTimes(1);

  expect(el).toMatchSnapshot();
});

test("button click", () => {
  render(<Button>Test button</Button>);

  const el = screen.getByTestId("button");

  expect(el).toHaveClass("primary-button");
  expect(el).toMatchSnapshot();
});

test("renders button outline", () => {
  render(<Button mode="outline">Test outline button</Button>);

  const el = screen.getByTestId("button");

  expect(el).toHaveClass("outline-button");
  expect(el).toMatchSnapshot();
});

test("renders button icon", () => {
  render(<Button mode="icon">X</Button>);

  const el = screen.getByTestId("button");

  expect(el).toHaveClass("icon-button");
  expect(el).toMatchSnapshot();
});
