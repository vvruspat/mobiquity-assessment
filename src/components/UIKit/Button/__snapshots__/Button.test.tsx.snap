// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button click 1`] = `
<button
  class="button primary-button"
  data-testid="button"
>
  Test button
</button>
`;

exports[`renders button 1`] = `
<button
  class="button primary-button"
  data-testid="button"
>
  Test button
</button>
`;

exports[`renders button icon 1`] = `
<button
  class="button icon-button"
  data-testid="button"
>
  X
</button>
`;

exports[`renders button outline 1`] = `
<button
  class="button outline-button"
  data-testid="button"
>
  Test outline button
</button>
`;
